diff -ruN gcc-16.0.0/gcc/config/elfos.h gcc-16.0.0-kos/gcc/config/elfos.h
--- gcc-16.0.0/gcc/config/elfos.h	2026-04-01 16:42:42.950821041 +0200
+++ gcc-16.0.0-kos/gcc/config/elfos.h	2026-04-01 16:59:01.923993708 +0200
@@ -486,3 +486,6 @@
 
 #undef TARGET_LIBC_HAS_FUNCTION
 #define TARGET_LIBC_HAS_FUNCTION no_c99_libc_has_function
+
+#define TARGET_OS_CPP_BUILTINS()			\
+  builtin_define ("__KOS_GCC_PATCHLEVEL__=2025062800")
diff -ruN gcc-16.0.0/gcc/config/sh/sh.md gcc-16.0.0-kos/gcc/config/sh/sh.md
--- gcc-16.0.0/gcc/config/sh/sh.md	2026-04-01 16:42:42.991949914 +0200
+++ gcc-16.0.0-kos/gcc/config/sh/sh.md	2026-04-01 17:09:39.819267734 +0200
@@ -6540,7 +6540,12 @@
 	      &&  ! satisfies_constraint_G (operands[1])
 	      &&  ! satisfies_constraint_H (operands[1])
 	      && REG_P (operands[0]))
-	    emit_insn (gen_movsf_ie_F_z (operands[0], operands[1]));
+	    {
+	      if (lra_in_progress)
+		emit_insn (gen_movsf_ie (operands[0], operands[1]));
+	      else
+		emit_insn (gen_movsf_ie_F_z (operands[0], operands[1]));
+	    }
 	  else if (REG_P (operands[0]) && REGNO (operands[0]) == FPUL_REG
 		   && satisfies_constraint_Q (operands[1]))
 	    emit_insn (gen_movsf_ie_Q_z (operands[0], operands[1]));
diff -ruN gcc-16.0.0/gcc/configure gcc-16.0.0-kos/gcc/configure
--- gcc-16.0.0/gcc/configure	2026-04-01 16:42:42.997726226 +0200
+++ gcc-16.0.0-kos/gcc/configure	2026-04-01 16:59:01.925023139 +0200
@@ -13218,7 +13218,7 @@
     target_thread_file='single'
     ;;
   aix | dce | lynx | mipssde | posix | rtems | \
-  single | tpf | vxworks | win32 | mcf)
+  single | tpf | vxworks | win32 | kos | mcf)
     target_thread_file=${enable_threads}
     ;;
   *)
diff -ruN gcc-16.0.0/libgcc/config/t-kos gcc-16.0.0-kos/libgcc/config/t-kos
--- gcc-16.0.0/libgcc/config/t-kos	1970-01-01 01:00:00.000000000 +0100
+++ gcc-16.0.0-kos/libgcc/config/t-kos	2026-04-01 16:59:01.926215842 +0200
@@ -0,0 +1 @@
+LIB2ADD = $(srcdir)/config/fake-kos.c
diff -ruN gcc-16.0.0/libgcc/config.host gcc-16.0.0-kos/libgcc/config.host
--- gcc-16.0.0/libgcc/config.host	2026-04-01 16:42:44.416583026 +0200
+++ gcc-16.0.0-kos/libgcc/config.host	2026-04-01 16:59:01.926107571 +0200
@@ -71,7 +71,7 @@
 asm_hidden_op=.hidden
 enable_execute_stack=
 extra_parts=
-tmake_file=
+tmake_file="t-kos"
 tm_file=
 tm_define=
 md_unwind_def_header=no-unwind.h
diff -ruN gcc-16.0.0/libgcc/configure gcc-16.0.0-kos/libgcc/configure
--- gcc-16.0.0/libgcc/configure	2026-04-01 16:42:44.443087652 +0200
+++ gcc-16.0.0-kos/libgcc/configure	2026-04-01 16:59:01.926391244 +0200
@@ -5826,6 +5826,7 @@
     tpf)	thread_header=config/s390/gthr-tpf.h ;;
     vxworks)	thread_header=config/gthr-vxworks.h ;;
     win32)	thread_header=config/i386/gthr-win32.h ;;
+    kos)	thread_header=gthr-kos.h ;;
     mcf)	thread_header=config/i386/gthr-mcf.h ;;
     *)
         as_fn_error $? "No known header for threading model '$target_thread_file'." "$LINENO" 5
diff -ruN gcc-16.0.0/libobjc/Makefile.in gcc-16.0.0-kos/libobjc/Makefile.in
--- gcc-16.0.0/libobjc/Makefile.in	2026-04-01 16:42:44.621337750 +0200
+++ gcc-16.0.0-kos/libobjc/Makefile.in	2026-04-01 16:59:01.926608986 +0200
@@ -308,14 +308,16 @@
 $(srcdir)/aclocal.m4: @MAINT@ $(aclocal_deps)
 	cd $(srcdir) && $(ACLOCAL) $(ACLOCAL_AMFLAGS)
 
-install: install-libs install-headers
+install-strip: INSTALL_STRIP_FLAG = -s
+install install-strip: install-libs install-headers
 
 install-libs: installdirs
 	$(SHELL) $(multi_basedir)/mkinstalldirs $(DESTDIR)$(toolexeclibdir)
-	$(LIBTOOL_INSTALL) $(INSTALL) libobjc$(libsuffix).la $(DESTDIR)$(toolexeclibdir);
+	$(LIBTOOL_INSTALL) $(INSTALL) $(INSTALL_STRIP_FLAG) \
+	  libobjc$(libsuffix).la $(DESTDIR)$(toolexeclibdir);
 	if [ "$(OBJC_BOEHM_GC)" ]; then \
-	  $(LIBTOOL_INSTALL) $(INSTALL) libobjc_gc$(libsuffix).la \
-				$(DESTDIR)$(toolexeclibdir);\
+	  $(LIBTOOL_INSTALL) $(INSTALL) $(INSTALL_STRIP_FLAG) \
+	    libobjc_gc$(libsuffix).la $(DESTDIR)$(toolexeclibdir);\
 	fi
 	$(MULTIDO) $(FLAGS_TO_PASS) multi-do DO="$@"
 	@-$(LIBTOOL) --mode=finish $(DESTDIR)$(toolexeclibdir)
@@ -328,7 +330,7 @@
 	  $(INSTALL_DATA) $${realfile} $(DESTDIR)$(libsubdir)/$(includedirname)/objc; \
 	done
 
-check uninstall install-strip dist installcheck installdirs:
+check uninstall dist installcheck installdirs:
 
 mostlyclean:
 	-$(LIBTOOL_CLEAN) rm -f libobjc$(libsuffix).la libobjc_gc$(libsuffix).la *.lo
diff -ruN gcc-16.0.0/libstdc++-v3/config/cpu/sh/atomicity.h gcc-16.0.0-kos/libstdc++-v3/config/cpu/sh/atomicity.h
--- gcc-16.0.0/libstdc++-v3/config/cpu/sh/atomicity.h	2026-04-01 16:42:44.675273171 +0200
+++ gcc-16.0.0-kos/libstdc++-v3/config/cpu/sh/atomicity.h	2026-04-01 16:59:01.926663637 +0200
@@ -22,14 +22,40 @@
 // see the files COPYING3 and COPYING.RUNTIME respectively.  If not, see
 // <http://www.gnu.org/licenses/>.
 
-// Use the default atomicity stuff, which will use __atomic* builtins
-// if threads are available, or the *_single functions on single-thread
-// configurations.
-// Actually we wouldn't need this header at all, but because of PR 53579
-// libstdc++'s configury will not pickup the -matomic-model= option when
-// set in the environment.  This makes it impossible to enable the proper
-// atomic model on SH without modifying GCC itself, because libstdc++ always
-// thinks the target doesn't do any atomics and uses the default mutex based
-// implementation from cpu/generic/atomicity_mutex.
+/* This is generic/atomicity.h */
 
 #include <ext/atomicity.h>
+#include <ext/concurrence.h>
+
+namespace 
+{
+  __gnu_cxx::__mutex&
+  get_atomic_mutex()
+  {
+    static __gnu_cxx::__mutex atomic_mutex;
+    return atomic_mutex;
+  }
+} // anonymous namespace
+
+namespace __gnu_cxx _GLIBCXX_VISIBILITY(default)
+{
+_GLIBCXX_BEGIN_NAMESPACE_VERSION
+
+  _Atomic_word
+  __attribute__ ((__unused__))
+  __exchange_and_add(volatile _Atomic_word* __mem, int __val) throw ()
+  {
+    __gnu_cxx::__scoped_lock sentry(get_atomic_mutex());
+    _Atomic_word __result;
+    __result = *__mem;
+    *__mem += __val;
+    return __result;
+  }
+
+  void
+  __attribute__ ((__unused__))
+  __atomic_add(volatile _Atomic_word* __mem, int __val) throw ()
+  { __exchange_and_add(__mem, __val); }
+
+_GLIBCXX_END_NAMESPACE_VERSION
+} // namespace
diff -ruN gcc-16.0.0/libstdc++-v3/configure gcc-16.0.0-kos/libstdc++-v3/configure
--- gcc-16.0.0/libstdc++-v3/configure	2026-04-01 16:42:44.677036740 +0200
+++ gcc-16.0.0-kos/libstdc++-v3/configure	2026-04-01 16:59:01.927802329 +0200
@@ -16391,6 +16391,7 @@
     tpf)	thread_header=config/s390/gthr-tpf.h ;;
     vxworks)	thread_header=config/gthr-vxworks.h ;;
     win32)	thread_header=config/i386/gthr-win32.h ;;
+    kos)	thread_header=gthr-kos.h ;;
     mcf)	thread_header=config/i386/gthr-mcf.h ;;
     *)
         as_fn_error $? "No known header for threading model '$target_thread_file'." "$LINENO" 5
